src/risclog/sqlalchemy/cache.py

```diff
@@ -0,0 +1,495 @@
+                if len(new_objects) > 0:
+                    sequence_values = self.session.using_bind(
+                        self._engine_name
+                    ).execute(
+                        "select nextval('%s') from generate_series(1,%s)"
+                        % (sequence[1], len(new_objects))
+                    )
+
+                    for object, value in zip(new_objects, sequence_values):
+                        setattr(object, sequence[0], value[0])
+
+    def _filter_sa_result_objects(self, objects):
+        return list(filter(lambda x: isinstance(x, ObjectBase), objects))
+
+    def _sync_relationship_attrs(self):
+        """
+        Synchronize relationship attributes with their corresponding ID
+        attributes.
+
+        When using SQLAlchemy bulk functionality, the ORM behavior is changed
+        in multiple ways. One important change is that relationship attributes
+        aren't evaluated anymore.
+        To still handle related object as expected, we iterate over
+        relationship attributes and set their corresponding ID attributes on
+        the same object.
+        """
+        for objects in self._cached_instances.values():
+            objects = self._filter_sa_result_objects(objects)
+            if len(objects) == 0:
+                continue
+            model = type(list(objects)[0])
+
+            rel_map = [
+                (rel.key, pair[1].key, pair[0].key)
+                for rel in inspect(model).relationships
+                for pair in rel.local_remote_pairs
+                if rel.direction is not sqlalchemy.orm.interfaces.ONETOMANY
+            ]
+            for object in objects:
+                for from_obj, from_attr, to_attr in rel_map:
+                    if getattr(object, to_attr, None) is None:
+                        setattr(
+                            object,
+                            to_attr,
+                            getattr(
+                                getattr(object, from_obj), from_attr, None
+                            ),
+                        )
+
+    def _register_change_handler(self, model, event_handler):
+        """Register an event handler for every column of a model."""
+        for attr in inspect(model).column_attrs:
+            sqlalchemy.event.listen(attr, 'set', event_handler)
+
+    def _deregister_change_handler(self, model, event_handler):
+        """Removes an event handler for every column of a model."""
+        for attr in inspect(model).column_attrs:
+            sqlalchemy.event.listen(attr, 'set', event_handler)
+
+    def _instance_change_handler(self, instance, value, oldvalue, initiator):
+        """
+        Re-index model instances that were changed
+        (to keep the index up-to-date).
+        Called by SQLAlchemy's model `set` event which fires when a model
+        attribute was changed. For more information,
+        see: https://docs.sqlalchemy.org/en/13/orm/events.html
+        """  # noqa: E501
```

Comment:
```suggestion
        """
```
Does not look as if it is necessary.